File: src/chain-observers/providers/internal-market-provider.ts
import { price, type asset } from "@hiveio/wax";
import { WorkerBeeIterable } from "../../types/iterator";
import { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import { OperationClassifier, IOperationTransactionPair } from "../classifiers/operation-classifier";
import { DataEvaluationContext } from "../factories/data-evaluation-context";
import { ProviderBase } from "./provider-base";

export interface IInternalMarketOperationBase {
  owner: string;
  orderId: number;
}

export interface IInternalMarketCancelOperation extends IInternalMarketOperationBase {
  cancel: true;
}

export interface IInternalMarketCreateOperation extends IInternalMarketOperationBase {
  cancel: false;
  amountToSell: asset;
  filled: boolean;
  exchangeRate: price;
  expiration: Date;
}

export type TInternalMarketOperation = IInternalMarketCancelOperation | IInternalMarketCreateOperation;

export interface IInternalMarketProviderData {
  internalMarketOperations: WorkerBeeIterable<IOperationTransactionPair<TInternalMarketOperation>>;
};

export class InternalMarketProvider extends ProviderBase {
  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [
      OperationClassifier
    ]
  }

  public async provide(data: DataEvaluationContext): Promise<IInternalMarketProviderData> {
    const operations = await data.get(OperationClassifier);

    const orderCreate = operations.operationsPerType.limit_order_create_operation;
    const orderCreate2 = operations.operationsPerType.limit_order_create2_operation;
    const orderCancel = operations.operationsPerType.limit_order_cancel_operation;

    const internalMarketOperations: IOperationTransactionPair<TInternalMarketOperation>[] = [];

    if(orderCancel)
      for(const op of orderCancel)
        internalMarketOperations.push({
          operation: {
            cancel: true,
            owner: op.operation.owner,
            orderId: op.operation.orderid
          },
          transaction: op.transaction
        });

    if(orderCreate)
      for(const op of orderCreate)
        internalMarketOperations.push({
          operation: {
            cancel: false,
            owner: op.operation.owner,
            orderId: op.operation.orderid,
            amountToSell: op.operation.amount_to_sell!,
            filled: op.operation.fill_or_kill,
            exchangeRate: {
              base: op.operation.amount_to_sell,
              quote: op.operation.min_to_receive
            },
            expiration: new Date(`${op.operation.expiration}Z`)
          },
          transaction: op.transaction
        });

    if(orderCreate2)
      for(const op of orderCreate2)
        internalMarketOperations.push({
          operation: {
            cancel: false,
            owner: op.operation.owner,
            orderId: op.operation.orderid,
            amountToSell: op.operation.amount_to_sell!,
            filled: op.operation.fill_or_kill,
            exchangeRate: op.operation.exchange_rate!,
            expiration: new Date(`${op.operation.expiration}Z`)
          },
          transaction: op.transaction
        });

    return {
      internalMarketOperations: new WorkerBeeIterable(internalMarketOperations)
    };
  }
}

File: python/workerbee/chain_observers/providers/internal_market_provider.py
"""InternalMarketProvider — provides internal market (limit order) operations.

Mirrors src/chain-observers/providers/internal-market-provider.ts.
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import TYPE_CHECKING

from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
from ..classifiers.operation_classifier import OperationClassifier
from .provider_base import ProviderBase

if TYPE_CHECKING:
    from ..factories.data_evaluation_context import DataEvaluationContext
    from ..payloads import InternalMarketPair, InternalMarketPayload


class InternalMarketProvider(ProviderBase):
    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        return [OperationClassifier]

    async def provide(self, data: DataEvaluationContext) -> InternalMarketPayload:
        operations = await data.get(OperationClassifier)
        ops_per_type = operations["operations_per_type"]

        order_create = ops_per_type.get("limit_order_create_operation")
        order_create2 = ops_per_type.get("limit_order_create2_operation")
        order_cancel = ops_per_type.get("limit_order_cancel_operation")

        internal_market_operations: list[InternalMarketPair] = []

        if order_cancel:
            for op in order_cancel:
                operation = op["operation"]
                owner = operation.get("owner")
                order_id = operation.get("orderid")
                if not isinstance(owner, str) or not isinstance(order_id, int):
                    continue
                internal_market_operations.append(
                    {
                        "operation": {
                            "cancel": True,
                            "owner": owner,
                            "order_id": order_id,
                        },
                        "transaction": op["transaction"],
                    }
                )

        if order_create:
            for op in order_create:
                operation = op["operation"]
                owner = operation.get("owner")
                order_id = operation.get("orderid")
                filled = operation.get("fill_or_kill")
                expiration_raw = operation.get("expiration")
                if not isinstance(owner, str) or not isinstance(order_id, int) or not isinstance(filled, bool):
                    continue
                if isinstance(expiration_raw, str):
                    expiration = datetime.fromisoformat(
                        expiration_raw if expiration_raw.endswith("Z") or "+" in expiration_raw else expiration_raw + "Z",
                    ).replace(tzinfo=UTC)
                elif isinstance(expiration_raw, datetime):
                    expiration = expiration_raw
                else:
                    continue

                internal_market_operations.append(
                    {
                        "operation": {
                            "cancel": False,
                            "owner": owner,
                            "order_id": order_id,
                            "amount_to_sell": operation.get("amount_to_sell"),
                            "filled": filled,
                            "exchange_rate": {
                                "base": operation.get("amount_to_sell"),
                                "quote": operation.get("min_to_receive"),
                            },
                            "expiration": expiration,
                        },
                        "transaction": op["transaction"],
                    }
                )

        if order_create2:
            for op in order_create2:
                operation = op["operation"]
                owner = operation.get("owner")
                order_id = operation.get("orderid")
                filled = operation.get("fill_or_kill")
                exchange_rate = operation.get("exchange_rate")
                expiration_raw = operation.get("expiration")
                if not isinstance(owner, str) or not isinstance(order_id, int) or not isinstance(filled, bool) or not isinstance(exchange_rate, dict):
                    continue
                if isinstance(expiration_raw, str):
                    expiration = datetime.fromisoformat(
                        expiration_raw if expiration_raw.endswith("Z") or "+" in expiration_raw else expiration_raw + "Z",
                    ).replace(tzinfo=UTC)
                elif isinstance(expiration_raw, datetime):
                    expiration = expiration_raw
                else:
                    continue

                internal_market_operations.append(
                    {
                        "operation": {
                            "cancel": False,
                            "owner": owner,
                            "order_id": order_id,
                            "amount_to_sell": operation.get("amount_to_sell"),
                            "filled": filled,
                            "exchange_rate": dict(exchange_rate),
                            "expiration": expiration,
                        },
                        "transaction": op["transaction"],
                    }
                )

        return {"internal_market_operations": internal_market_operations}
